### PR TITLE
feat(FX-3631): Add animation for the artwork grid filter menu header

### PR DESCRIPTION
**Pull request by dimatretyak**
The type of this PR is: **Feature**

This PR resolves [FX-3631]

⚠️ Changes behind `AREnableImprovedAlertsFlow` feature flag

### Acceptance Criteria
* Animation is applied during scrolling of content
* Final state of animation:
  * The title is on the same level as the "back arrow" icon
  * The size of the title text is `md`
* Try to make a component that can be reused

### Demo
|State|iOS|Android|
|-|-|-|
|Initial|Simulator Screen Shot - iPhone 8 - 2021-12-13 at 16 56 46|photo_2021-12-13 16 58 59|
|Final|Simulator Screen Shot - iPhone 8 - 2021-12-13 at 16 56 50|photo_2021-12-13 16 58 57|

#### iOS
https://user-images.githubusercontent.com/3513494/145823888-5eb3b0ef-736c-4b0a-b5d5-9cb621fd2803.mp4

#### Android
https://user-images.githubusercontent.com/3513494/145823918-9ef1d387-6994-49ce-805c-2caa38320a9d.mp4

### PR Checklist (tick all before merging)

- [x] I have included screenshots or videos to illustrate my changes, or I have not changed anything that impacts the UI.
- [x] I have tested my changes on **iOS** and **Android**.
- [x] I have added tests/stories for my changes, or my changes don't require testing/stories, or I have included a link to a separate Jira ticket covering the tests.
- [x] I have added a feature flag, or my changes don't require a feature flag. ([How do I add one?](https://github.com/artsy/eigen/blob/master/docs/developing_a_feature.md))
- [x] I have documented any follow-up work that this PR will require, or it does not require any.
- [x] I have added an app state migration, or my changes do not require one. ([What are migrations?](https://github.com/artsy/eigen/blob/master/docs/adding_state_migrations.md))
- [x] I have added a changelog entry below or my changes do not require one.

### To the reviewers 👀

- [x] I would like at least one of the reviewers to run this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

- Add animation for the artwork grid filter menu header - dimatretyak

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

</details>

[FX-3631]: https://artsyproduct.atlassian.net/browse/FX-3631?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
### This PR contains the following changes:

- Cross-platform user-facing changes (Add animation for the artwork grid filter menu header - dimatretyak)
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against a99c27d1ae47adb44b9b313acd279ace7a3e1b22
</p>

File: src/lib/Components/AnimatableHeader/AnimatableHeader.tsx [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        style={{
+          flex: 1,
+          justifyContent: "center",
+          height: headerHeight,
+          marginLeft: space(0.5) + space(1),
+          opacity,
+          transform: [{ translateX }],
+        }}
+      >
+        <Text numberOfLines={2} lineHeight={18}>
```
Comment (MounirDhahri) [inline src/lib/Components/AnimatableHeader/AnimatableHeader.tsx:125]: **praise:** nice

File: src/lib/Components/AnimatableHeader/AnimatableHeaderLargeTitle.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  const isEnabledImprovedAlertsFlow = useFeatureFlag("AREnableImprovedAlertsFlow")
+  const space = useSpace()
+  const { scrollOffsetY, largeTitleVerticalOffset, largeTitleHeight, largeTitleEndEdge, title } =
+    useAnimatableHeaderContext()
+
+  if (isEnabledImprovedAlertsFlow) {
+    const largeTitleOpacity = Animated.interpolate(scrollOffsetY, {
+      inputRange: [0, largeTitleEndEdge],
+      outputRange: [1, 0],
+      extrapolate: Extrapolate.CLAMP,
```
Comment (MounirDhahri) [inline src/lib/Components/AnimatableHeader/AnimatableHeaderLargeTitle.tsx:17]: **praise:** good choice

**Review by MounirDhahri — DISMISSED** [review]
Really nice!!! great work
I left few comments but they're not blocking

File: src/lib/Components/AnimatableHeader/AnimatableHeaderScrollView.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import React from "react"
+import { ScrollView, ScrollViewProps } from "react-native"
+import Animated from "react-native-reanimated"
+import { useAnimatableHeaderContext } from "./AnimatableHeaderContext"
+import { AnimatableHeaderLargeTitle } from "./AnimatableHeaderLargeTitle"
+
+const AnimatedComponent: typeof ScrollView = Animated.createAnimatedComponent(ScrollView)
```
Comment (gkartalis) [inline src/lib/Components/AnimatableHeader/AnimatableHeaderScrollView.tsx:8]: **nitpick(non-blocking):** Maybe AnimatedScrollView would be a better name here? Up to you though to change that or not

File: src/lib/Components/AnimatableHeader/AnimatableHeader.tsx [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+      set(config.toValue, 0),
+      startClock(clock),
+    ]),
+    // we run the step here that is going to update position
+    timing(clock, state, config),
+    cond(state.finished, stopClock(clock)),
+    // we made the block return the updated position
+    state.position,
+  ])
+}
```
Comment (gkartalis) [inline src/lib/Components/AnimatableHeader/AnimatableHeader.tsx:69]: **praise:** Amazingly smooth animation! Had to revisit reanimated docs thought to understand it. 
Maybe in the **future** we want to move to using reanimated v2 in eigen that has a more human friendly and understandable api.

File: src/lib/Components/AnimatableHeader/AnimatableHeaderContext.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+export const AnimatableHeaderContext = React.createContext<AnimatableHeaderContextType>({
+  scrollOffsetY: new Animated.Value(0),
+  headerHeight: 0,
+  largeTitleVerticalOffset: 0,
+  title: "",
+  largeTitleHeight: new Animated.Value(0),
+  largeTitleEndEdge: new Animated.Value(0),
+  // tslint:disable-next-line:no-empty
+  setTitle: () => {},
```
[the remark this answers is not in the document]
Comment (gkartalis) [reply src/lib/Components/AnimatableHeader/AnimatableHeaderContext.ts:22]: **thought:** I think that I wouldn't move that to `easy-peasy` not sure about how `useReducer` would be beneficial here either 🤔

File: src/lib/Components/AnimatableHeader/AnimatableHeaderFlatList.tsx [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+const AnimatedComponent: typeof FlatList = Animated.createAnimatedComponent(FlatList)
+
+export const AnimatableHeaderFlatList = <T extends any>(props: FlatListProps<T>) => {
+  const { ListHeaderComponent, ...other } = props
+  const { scrollOffsetY } = useAnimatableHeaderContext()
+
+  return (
+    <AnimatedComponent
+      {...other}
+      scrollEventThrottle={0.0000000001}
```
Comment (gkartalis) [inline src/lib/Components/AnimatableHeader/AnimatableHeaderFlatList.tsx]: **question(non-blocking):** Why do we need this value here? Looks really specific to me. The [docs](https://reactnative.dev/docs/scrollview#scrolleventthrottle-ios) mention that _"The default value is 0, which results in the scroll event being sent only once each time the view is scrolled."_ Wouldn't this value result in  more information being sent across the bridge?

**Review by gkartalis — APPROVED** [review]
Looks great to me and runs super smooth in emulators as well!  There are some tests failing though. Added some questions as well but other than that great work! 🚀

File: src/lib/Components/AnimatableHeader/AnimatableHeader.tsx [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+      >
+        <ArrowLeftIcon fill="black100" />
+      </TouchableOpacity>
+      <Animated.View
+        pointerEvents="none"
+        style={{
+          flex: 1,
+          justifyContent: "center",
+          height: headerHeight,
+          marginLeft: space(0.5) + space(1),
```
[the remark this answers is not in the document]
Comment (dimatretyak) [reply src/lib/Components/AnimatableHeader/AnimatableHeader.tsx:120]: Unfortunately, if we use `space(1.5)`, there will be the same offset as if you use `space(2)`

bug

File: src/lib/Components/AnimatableHeader/AnimatableHeaderScrollView.tsx (hunk 3 above)
Comment (dimatretyak) [reply src/lib/Components/AnimatableHeader/AnimatableHeaderScrollView.tsx:8]: Good suggestion. Thanks!

File: src/lib/Components/AnimatableHeader/AnimatableHeader.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (gkartalis) [reply src/lib/Components/AnimatableHeader/AnimatableHeader.tsx:120]: gotcha, thanks for clarifying

File: src/lib/Components/AnimatableHeader/AnimatableHeaderContext.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (dimatretyak) [reply src/lib/Components/AnimatableHeader/AnimatableHeaderContext.ts:22]: > What about using useReducer or component store from easy-peasy?

Can you describe the idea of using `easy-peasy` or `useReducer` in a little more detail here?

The main idea of using `setTitle` in context is that the user does not have to specify the same prop 2 times for `AnimatableHeader` and `AnimatableHeaderFlatList`/`AnimatableHeaderScrollView`.

File: src/lib/Components/AnimatableHeader/AnimatableHeaderFlatList.tsx (hunk 6 above)
Comment (dimatretyak) [reply src/lib/Components/AnimatableHeader/AnimatableHeaderFlatList.tsx]: I decided to set the same value as it is set in [StickyHeaderPage](https://github.com/artsy/eigen/blob/70fb25078d135db4ed4d247e216be5a6e5103c19/src/lib/Components/StickyHeaderPage.tsx#L67), [InfiniteScrollArtworksGrid](https://github.com/artsy/eigen/blob/70fb25078d135db4ed4d247e216be5a6e5103c19/src/lib/Components/ArtworkGrids/InfiniteScrollArtworksGrid.tsx#L365), [ImageZoomView](https://github.com/artsy/eigen/blob/70fb25078d135db4ed4d247e216be5a6e5103c19/src/lib/Scenes/Artwork/Components/ImageCarousel/FullScreen/ImageZoomView.tsx#L305), [VerticalSwipeToDismiss](https://github.com/artsy/eigen/blob/70fb25078d135db4ed4d247e216be5a6e5103c19/src/lib/Scenes/Artwork/Components/ImageCarousel/FullScreen/VerticalSwipeToDismiss.tsx#L105)

File: src/lib/Components/AnimatableHeader/AnimatableHeaderContext.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (MounirDhahri) [reply src/lib/Components/AnimatableHeader/AnimatableHeaderContext.ts:22]: I am fine keeping it this way. In this context maybe this dependency inversion might not be harmful as I don't see us using/setting the setTitle from anywhere else but I wanted to flag this so we're aware of it.

**Pull request merged**